### PR TITLE
os-tor Extra settings for source IP and IPv6

### PR DESCRIPTION
**Pull request by gizahNL**
Would partially solve:
https://github.com/opnsense/plugins/issues/531

 Quick fixes to allow IPv6 and Outboundbinding:

-allow specification of secondary ORPort (intended for IPv6)
-allow specification of 2 source addresses via source IP fields,
especially usefull for exits.
-default directory port to 9030 and set to mandatory (needed for relay)

All of the IPv6 fields use the same input verification as the HOST
field, should perhaps be changed?

### CODE REVIEW

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 1]
```diff
@@ -10,6 +10,18 @@
       <Required>N</Required>
       <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-f0-9:]{2,})$/i</mask>
     </host>
+    <hostv6 type="TextField">
+      <Required>N</Required>
+      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-f0-9:]{2,})$/i</mask>
+    </hostv6>
+    <outboundbind type="TextField">
+      <Required>N</Required>
+      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-f0-9:]{2,})$/i</mask>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: ``|[a-f0-9:]{2,}`` is for v6 and makes no sense here as well as v4 at the bottom

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 2]
```diff
@@ -19,7 +31,8 @@
     </port>
     <directory_port type="IntegerField">
       <MinimumValue>1</MinimumValue>
-      <Required>N</Required>
+      <Required>Y</Required>
+      <default>9030</default>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: isn't that producing much traffic?

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 3]
```diff
@@ -171,14 +188,14 @@ RelayBandwidthBurst {{ OPNsense.tor.relay.bandwithburst }} KBytes
...
-{% endif %}
+DirPort {% if helpers.exists('OPNsense.tor.relay.host') and OPNsense.tor.relay.host != '' %}{{ OPNsense.tor.relay.host }}:{% endif%}{{ OPNsense.tor.relay.directory_port|default('9030') }}
+
 ## Uncomment to return an arbitrary blob of html on your DirPort. Now you
 ## can explain what Tor is if anybody wonders why your IP address is
 ## contacting them. See contrib/tor-exit-notice.html in Tor's source
 ## distribution for a sample.
-#DirPortFrontPage /usr/local/etc/tor/tor-exit-notice.html
+
+DirPortFrontPage /usr/local/etc/tor/tor-exit-notice.html
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: nonsense without being able to set the file content.

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 4]
```diff
@@ -203,6 +220,9 @@ ExitPolicy {{ policy.action }}{% if policy.type == 'v6' %}6{% endif
 {%   endif %}
 {% endif %}
 
+%include /usr/local/etc/torrc.exitpolicy
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: the exit policy is already configurable via the GUI

**Comment by gizahNL** [comment]
Made some changes.

The include bit was not intended to be for merge, it is there for me to locally include it, same for DirFrontPage.
Though the Frontpage should be possible to configured, and upload or paste some flat HTML. 

And the current exit settings are completely broken, it is impossible to manage a complex list with wildcard adresses, port ranges without incredible micro. 
I'm not even sure if it is possible to enter wildcard adresses in the exit ACL

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 5]
```diff
@@ -8,8 +8,20 @@
     </enabled>
     <host type="TextField">
       <Required>N</Required>
-      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-f0-9:]{2,})$/i</mask>
+      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}</mask>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: `$/` is missing - shouldn't that thow an exception on your machine? That's an invalid regular expression

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 6]
```diff
@@ -128,6 +128,12 @@ HiddenServicePort {{ acl.port}} {{ acl.target_host }}:{{ acl.target_port }}
 
 ORPort {% if helpers.exists('OPNsense.tor.relay.host') and OPNsense.tor.relay.host != '' %}{{ OPNsense.tor.relay.host }}:{% endif%}{{ OPNsense.tor.relay.port|default('9001') }}
 
+{% if helpers.exists('OPNsense.tor.relay.hostv6') and OPNsense.tor.relay.hostv6 != '' %}
+
+ORPort [{{ OPNsense.tor.relay.hostv6 }}]:{{ OPNsense.tor.relay.port|default('9001') }}
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: remove the empty line before and after this line like in the other options

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 7]
```diff
@@ -171,13 +188,13 @@ RelayBandwidthBurst {{ OPNsense.tor.relay.bandwithburst }} KBytes
 
 #ContactInfo 0xFFFFFFFF Random Person <nobody AT example dot com>
 
-{% if helpers.exists('OPNsense.tor.relay.directory_port') and OPNsense.tor.relay.directory_port != '' %}
-DirPort {{ OPNsense.tor.relay.directory_port }}
-{% endif %}
+DirPort {% if helpers.exists('OPNsense.tor.relay.host') and OPNsense.tor.relay.host != '' %}{{ OPNsense.tor.relay.host }}:{% endif%}{{ OPNsense.tor.relay.directory_port|default('9030') }}
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: This should be still disabled (not set) by default. Restore the original check and add the host to the original DirPort line

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 8]
```diff
@@ -203,6 +220,8 @@ ExitPolicy {{ policy.action }}{% if policy.type == 'v6' %}6{% endif
 {%   endif %}
 {% endif %}
 
+
+
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: remove those extra empty lines

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 9]
```diff
@@ -213,3 +232,5 @@ PublishServerDescriptor {{ OPNsense.tor.relay.publish|default('0') }}
 {% endif %}
 
 {% endif %}
+
+
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: same here

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 10]
```diff
@@ -8,8 +8,20 @@
     </enabled>
     <host type="TextField">
       <Required>N</Required>
-      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-f0-9:]{2,})$/i</mask>
+      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/i</mask>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: bracket issue (remove the opening or add a closing bracket)

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 11]
```diff
@@ -8,8 +8,20 @@
...
     </host>
+    <hostv6 type="TextField">
+      <Required>N</Required>
+      <mask>/^([a-f0-9:]{2,})$/i</mask>
+    </hostv6>
+    <outboundbind type="TextField">
+      <Required>N</Required>
+      <mask>/^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/i</mask>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: same here

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml [hunk 12]
```diff
@@ -19,7 +31,7 @@
     </port>
     <directory_port type="IntegerField">
       <MinimumValue>1</MinimumValue>
-      <Required>N</Required>
+      <Required>n</Required>
```
Comment (fichtner) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/Relay.xml]: Let's keep "N" ?

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 13]
```diff
@@ -54,16 +54,14 @@ SOCKSPolicy {{ policy.action }}{% if policy.type == 'v6' %}6{% endif %} {{ polic
 SOCKSPolicy reject *
 SOCKSPolicy reject6 *
 
-
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: For better or worse, it's better to keep the extra lines than to push them in a feature addition

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 14]
```diff
@@ -62,8 +62,7 @@ Log {{ OPNsense.tor.general.logfilelevel }} file /var/log/tor.log
 Log {{ OPNsense.tor.general.sysloglevel }} syslog
 {% endif %}
 
-
-
+Schedulers KISTLite
```
Comment (fabianfrz) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: from the man page:
```
Schedulers KIST|KISTLite|Vanilla
```
So this should be a dropdown in the menu which is not requrired so if nothing is selected, the default value is used

**Comment by gizahNL** [comment]
It might've been a coincedence, however I noticed no traffic passing over my relay nor my tor getting published (It was publishing itself, however this was not reaching the consensus) before I set the scheduler to KISTLite, so it might be that the unset default (which defaults to KIST) is not working. 
So I would recommend defaulting to KISTLite, and giving a dropdown to select between KISTLite and Vanilla on general tab.

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 15]
```diff
@@ -62,7 +62,18 @@ Log {{ OPNsense.tor.general.logfilelevel }} file /var/log/tor.log
 Log {{ OPNsense.tor.general.sysloglevel }} syslog
 {% endif %}
 
-
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: retain whitespace here

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 16]
```diff
@@ -170,21 +190,21 @@ RelayBandwidthBurst {{ OPNsense.tor.relay.bandwithburst }} KBytes
...
 {% if helpers.exists('OPNsense.tor.relay.directory_port') and OPNsense.tor.relay.directory_port != '' %}
-DirPort {{ OPNsense.tor.relay.directory_port }}
+DirPort {% if helpers.exists('OPNsense.tor.relay.host') and OPNsense.tor.relay.host != '' %}{{ OPNsense.tor.relay.host }}:{% endif%}{{ OPNsense.tor.relay.directory_port }}
 {% endif %}
+
 ## Uncomment to return an arbitrary blob of html on your DirPort. Now you
 ## can explain what Tor is if anybody wonders why your IP address is
 ## contacting them. See contrib/tor-exit-notice.html in Tor's source
 ## distribution for a sample.
+
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: here

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 17]
```diff
@@ -170,21 +190,21 @@ RelayBandwidthBurst {{ OPNsense.tor.relay.bandwithburst }} KBytes
...
 #DirPortFrontPage /usr/local/etc/tor/tor-exit-notice.html
 
 {% if helpers.exists('OPNsense.tor.relay.exitrejectprivateip') %}
 ExitPolicyRejectPrivate {{ OPNsense.tor.relay.exitrejectprivateip }}
 {% endif %}
 
-
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: here

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 18]
```diff
@@ -207,7 +227,6 @@ ExitPolicy {{ policy.action }}{% if policy.type == 'v6' %}6{% endif
 ExitPolicy reject *:*
 ExitPolicy reject6 *:*
 
-
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: and here

File: security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc [hunk 19]
```diff
@@ -170,21 +190,21 @@ RelayBandwidthBurst {{ OPNsense.tor.relay.bandwithburst }} KBytes
 #AccountingStart month 3 15:00
 
 #ContactInfo 0xFFFFFFFF Random Person <nobody AT example dot com>
-
```
Comment (fichtner) [inline security/tor/src/opnsense/service/templates/OPNsense/Tor/torrc]: here too

File: security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/General.xml [hunk 20]
```diff
@@ -64,6 +64,17 @@
         <debug>Debugging</debug>
       </OptionValues>
     </sysloglevel>
+    <scheduler type="OptionField">
+      <Required>Y</Required>
+      <multiple>N</multiple>
+      <default>KISTLite,Vanilla</default>
```
Comment (fabianfrz) [inline security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/General.xml]: default must be an option from OptionValues below (remove the comma)

Comment (fabianfrz) [reply security/tor/src/opnsense/mvc/app/models/OPNsense/Tor/General.xml]: (comma would be in use if it would be a multiple field)

**Comment by gizahNL** [comment]
please let me know if there are other things that need to be changed, I'll give em some attention somewhere this week. 
I'm planning to take some time for ExitACLs this weekend

**Comment by fabianfrz** [comment]
Looks good will bump the version so it can go to the next release.

**Comment by fabianfrz** [comment]
@gizahNL Would you like to update the docs (add the new options [here](https://github.com/opnsense/docs/blob/master/source/manual/how-tos/tor.rst#relays))?

**Comment by gizahNL** [comment]
ill put it on my todo list, ping me if it starts to look like I might have forgotten

**Pull request merged**